runtime/major_gc.c

```diff
@@ -702,40 +707,56 @@ static void mark_slice_darken(struct mark_stack* stk, value v, mlsize_t i,
   }
 }
 
-static intnat do_some_marking(intnat budget) {
-  struct mark_stack* stk = Caml_state->mark_stack;
+static intnat do_some_marking(struct mark_stack* stk, intnat budget) {
   while (stk->count > 0) {
     mark_entry me = stk->stack[--stk->count];
-    intnat me_end = Wosize_val(me.block);
-    while (me.offset != me_end) {
+    while (me.start < me.end) {
       if (budget <= 0) {
-        mark_stack_push(stk, me.block, me.offset, NULL);
+        mark_stack_push_range(stk, me.start, me.end);
         return budget;
       }
       budget--;
-      CAMLassert(Is_markable(me.block) &&
-                 Has_status_hd(Hd_val(me.block),
-                               caml_global_heap_state.MARKED) &&
-                 Tag_val(me.block) < No_scan_tag &&
-                 Tag_val(me.block) != Cont_tag);
-      mark_slice_darken(stk, me.block, me.offset++, &budget);
+      mark_slice_darken(stk, *me.start, &budget);
+      me.start++;
     }
     budget--; /* credit for header */
   }
   return budget;
 }
 
+/* compressed mark stack */
+#define PAGE_MASK (~(uintnat)(BITS_PER_WORD-1))
+#define PTR_TO_PAGE(v) (((uintnat)(v)>>3) & PAGE_MASK)
+#define PTR_TO_PAGE_OFFSET(v) ((((uintnat)(v)>>3) & ~PAGE_MASK))
```

Comment:
Is this really working on 32-bit machines? Shouldn't it be something like this?
```suggestion
#define PTR_TO_PAGE(v) (((uintnat)(v)/sizeof(value)) & PAGE_MASK)
#define PTR_TO_PAGE_OFFSET(v) ((((uintnat)(v)/sizeof(value)) & ~PAGE_MASK))
```

Reply:
Good catch: fixed f02eb61